src/test/owr/OptimisticWithdrawalRecipientV2.t.sol

```diff
@@ -0,0 +1,730 @@
+    vm.stopPrank();
+
+    bytes[] memory requestsMade = withdrawalMock.getRequests();
+    assertEq(requestsMade.length, numRequests);
+    assertEq(_user.balance, excessFee);
+    assertEq(address(withdrawalMock).balance, expectedTotalFee);
+    for (uint256 i; i < numRequests; i++) {
+      bytes memory requestData = abi.encodePacked(pubkeys[i], amounts[i]);
+      assertEq(requestsMade[i], requestData);
+    }
+  }
+
+  function testReceiveETH() public {
+    address(owrETH).safeTransferETH(1 ether);
+    assertEq(address(owrETH).balance, 1 ether);
+  }
+
+  function testReceiveTransfer() public {
+    payable(address(owrETH)).transfer(1 ether);
+    assertEq(address(owrETH).balance, 1 ether);
+  }
+
+  function testEmitOnReceiveETH() public {
+    vm.expectEmit(true, true, true, true);
+    emit ReceiveETH(1 ether);
+
+    address(owrETH).safeTransferETH(1 ether);
+  }
+
+  function testReceiveERC20() public {
+    address(mERC20).safeTransfer(address(owrETH), 1 ether);
+    assertEq(mERC20.balanceOf(address(owrETH)), 1 ether);
+  }
+
+  function testCan_recoverNonOWRFundsToRecipient() public {
+    address(owrETH).safeTransferETH(1 ether);
+    address(mERC20).safeTransfer(address(owrETH), 1 ether);
+    address(owrETH_OR).safeTransferETH(1 ether);
+    address(mERC20).safeTransfer(address(owrETH_OR), 1 ether);
+
+    vm.expectEmit(true, true, true, true);
+    emit RecoverNonOWRecipientFunds(address(mERC20), recoveryAddress, 1 ether);
+    owrETH.recoverFunds(address(mERC20), recoveryAddress);
+    assertEq(address(owrETH).balance, 1 ether);
+    assertEq(mERC20.balanceOf(address(owrETH)), 0 ether);
+    assertEq(mERC20.balanceOf(recoveryAddress), 1 ether);
+
+    vm.expectEmit(true, true, true, true);
+    emit RecoverNonOWRecipientFunds(address(mERC20), principalRecipient, 1 ether);
+    owrETH_OR.recoverFunds(address(mERC20), principalRecipient);
+    assertEq(address(owrETH_OR).balance, 1 ether);
+    assertEq(mERC20.balanceOf(address(owrETH_OR)), 0 ether);
+    assertEq(mERC20.balanceOf(principalRecipient), 1 ether);
+
+    address(mERC20).safeTransfer(address(owrETH_OR), 1 ether);
+
+    vm.expectEmit(true, true, true, true);
+    emit RecoverNonOWRecipientFunds(address(mERC20), rewardsRecipient, 1 ether);
+    owrETH_OR.recoverFunds(address(mERC20), rewardsRecipient);
+    assertEq(address(owrETH_OR).balance, 1 ether);
+    assertEq(mERC20.balanceOf(address(owrETH_OR)), 0 ether);
+    assertEq(mERC20.balanceOf(rewardsRecipient), 1 ether);
+  }
+
+  function testCannot_recoverFundsToNonRecipient() public {
+    vm.expectRevert(OptimisticWithdrawalRecipientV2.InvalidTokenRecovery_InvalidRecipient.selector);
+    owrETH.recoverFunds(address(mERC20), address(1));
+
+    vm.expectRevert(OptimisticWithdrawalRecipientV2.InvalidTokenRecovery_InvalidRecipient.selector);
+    owrETH_OR.recoverFunds(address(mERC20), address(2));
+  }
+
+  function testCan_OWRIsPayable() public {
+    owrETH.distributeFunds{value: 2 ether}();
```

Comment:
Weird, is this insta-adding and distributing msg.value? I guess it works? 🤷

Reply:
This part came from V1 test, I did not examine it in details.